dll/win32/shell32/CDefViewUtil.cpp

```diff
@@ -0,0 +1,132 @@
+/*
+ * PROJECT:     shell32
+ * LICENSE:     LGPL-2.1+ (https://spdx.org/licenses/LGPL-2.1+)
+ * PURPOSE:     DefView utility functions
+ * COPYRIGHT:   Copyright 2024 Whindmar Saksit <whindsaks@proton.me>
+ */
+
+#include "precomp.h"
+
+WINE_DEFAULT_DEBUG_CHANNEL(shell);
+
+class CObjectWithSiteBase :
+    public IObjectWithSite
+{
+    public:
+        IUnknown* m_pUnkSite;
+
+        CObjectWithSiteBase() : m_pUnkSite(NULL) {}
+        virtual ~CObjectWithSiteBase() { SetSite(NULL); }
+
+        // IObjectWithSite
+        STDMETHODIMP SetSite(IUnknown *pUnkSite) override
+        {
+            IUnknown_Set(&m_pUnkSite, pUnkSite);
+            return S_OK;
+        }
+        STDMETHODIMP GetSite(REFIID riid, void **ppvSite) override
+        {
+            *ppvSite = NULL;
+            return m_pUnkSite ? m_pUnkSite->QueryInterface(riid, ppvSite) : E_FAIL;
+        }
+};
+
+// This class adapts the legacy function callback to work as an IShellFolderViewCB
+class CShellFolderViewCBWrapper :
+    public CComObjectRootEx<CComMultiThreadModelNoCS>,
+    public IShellFolderViewCB,
+    public CObjectWithSiteBase
+{
+    protected:
+        HWND                    m_hWndMain;
+        CComPtr<IShellFolder>   m_psf;
+        CComPtr<IShellView>     m_psvOuter;
+        LPFNVIEWCALLBACK        m_Callback;
+        FOLDERVIEWMODE          m_FVM;
+        LONG                    m_Events;
+    public:
+        CShellFolderViewCBWrapper() : m_hWndMain(NULL) {}
+
+        HRESULT WINAPI Initialize(LPCSFV psvcbi)
+        {
+            m_psf = psvcbi->pshf;
+            m_psvOuter = psvcbi->psvOuter;
+            m_Callback = psvcbi->pfnCallback;
+            m_FVM = psvcbi->fvm;
+            m_Events = psvcbi->lEvents;
+            return S_OK;
+        }
+
+        // IShellFolderViewCB
+        STDMETHODIMP MessageSFVCB(UINT uMsg, WPARAM wParam, LPARAM lParam) override
+        {
+            switch (uMsg)
+            {
+                case SFVM_HWNDMAIN:
+                    m_hWndMain = (HWND)lParam;
+                    break;
+
+                case SFVM_DEFVIEWMODE:
+                    if (m_FVM)
+                        *(FOLDERVIEWMODE*)lParam = m_FVM;
+                    break;
```

Comment:
No pre-treatment needs to be done for some other SFVM_* messages?

Reply:
Probably but most SFVM messages are poorly or not at all documented.

Reply:
OK I guess we'll have to see on real cases in the future :thinking: